CPP/DesktopToastsCppApp/DesktopNotificationManagerCompat.cpp

```diff
@@ -0,0 +1,287 @@
+#include <appmodel.h>
+
+#define THROW_IF_FAILED(hr) if (FAILED(hr)) { throw hr; }
+
+using namespace ABI::Windows::Data::Xml::Dom;
+using namespace Microsoft::WRL;
+using namespace Microsoft::WRL::Wrappers;
+
+namespace DesktopNotificationManagerCompat
+{
+    void RegisterComServer(GUID clsid, const wchar_t exePath[]);
+    void EnsureRegistered();
+    bool IsRunningAsUwp();
+
+    bool s_registeredAumidAndComServer = false;
+    std::wstring s_aumid;
+    bool s_registeredActivator = false;
+    bool s_hasCheckedIsRunningAsUwp = false;
+    bool s_isRunningAsUwp = false;
+
+    void RegisterAumidAndComServer(const wchar_t *aumid, GUID clsid)
+    {
+        // If running as Desktop Bridge
+        if (IsRunningAsUwp())
+        {
+            // Clear the AUMID since Desktop Bridge doesn't use it, and then we're done.
+            // Desktop Bridge apps are registered with platform through their manifest.
+            // Their LocalServer32 key is also registered through their manifest.
+            s_aumid = L"";
+            s_registeredAumidAndComServer = true;
+            return;
+        }
+
+        // Copy the aumid
+        s_aumid = std::wstring(aumid);
+
+        // Get the EXE path
+        wchar_t exePath[MAX_PATH];
+        DWORD charWritten = ::GetModuleFileName(nullptr, exePath, ARRAYSIZE(exePath));
+        THROW_IF_FAILED(charWritten > 0 ? S_OK : HRESULT_FROM_WIN32(::GetLastError()));
+
+        // Register the COM server
+        RegisterComServer(clsid, exePath);
+        s_registeredAumidAndComServer = true;
+    }
+
+    void RegisterActivator()
+    {
+        // Module<OutOfProc> needs a callback registered before it can be used.
+        // Since we don't care about when it shuts down, we'll pass an empty lambda here.
+        Module<OutOfProc>::Create([] {});
+
+        // If a local server process only hosts the COM object then COM expects
+        // the COM server host to shutdown when the references drop to zero.
+        // Since the user might still be using the program after activating the notification,
+        // we don't want to shutdown immediately.  Incrementing the object count tells COM that
+        // we aren't done yet.
+        Module<OutOfProc>::GetModule().IncrementObjectCount();
+
+        THROW_IF_FAILED(Module<OutOfProc>::GetModule().RegisterObjects());
+
+        s_registeredActivator = true;
+    }
+
+    void RegisterComServer(GUID clsid, const wchar_t exePath[])
+    {
+        // Turn the GUID into a string
+        OLECHAR* clsidOlechar;
+        StringFromCLSID(clsid, &clsidOlechar);
+        std::wstring clsidStr(clsidOlechar);
+        ::CoTaskMemFree(clsidOlechar);
+
+        // Create the subkey
+        // Something like SOFTWARE\Classes\CLSID\{23A5B06E-20BB-4E7E-A0AC-6982ED6A6041}\LocalServer32
+        std::wstring subKeyStr = LR"(SOFTWARE\Classes\CLSID\)" + clsidStr + LR"(\LocalServer32)";
+        LPCWSTR subKey = subKeyStr.c_str();
```

Comment:
don't bother with this, just call subKeyStr.c_str() down below. Also, just call the std::string variable subkey.